luigi/contrib/lsf.py

```diff
@@ -0,0 +1,354 @@
+# -*- coding: utf-8 -*-
+
+"""
+.. Copyright 2012-2015 Spotify AB
+   Copyright 2018
+   Copyright 2018 EMBL-European Bioinformatics Institute
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+"""
+
+import os
+import subprocess
+import time
+import sys
+import logging
+import random
+import shutil
+try:
+    # Dill is used for handling pickling and unpickling if there is a deference
+    # in server setups between the LSF submission node and the nodes in the
+    # cluster
+    import dill as pickle
+except ImportError:
+    import pickle
+
+import luigi
+import luigi.configuration
+from luigi.contrib.hadoop import create_packages_archive
+from luigi.contrib import lsf_runner
+from luigi.task_status import PENDING, FAILED, DONE, RUNNING, UNKNOWN
+
+"""
+LSF batch system Tasks.
+=======================
+
+What's LSF? see http://en.wikipedia.org/wiki/Platform_LSF
+and https://wiki.med.harvard.edu/Orchestra/IntroductionToLSF
+
+See: https://github.com/spotify/luigi/issues/1936
+
+This extension is modeled after the hadoop.py approach.
+I'll be making a few assumptions, and will try to note them.
+
+Going into it, the assumptions are:
+
+- You schedule your jobs on an LSF submission node.
+- The 'bjobs' command on an LSF batch submission system returns a standardized format.
+- All nodes have access to the code you're running.
+- The sysadmin won't get pissed if we run a 'bjobs' check every thirty
+  seconds or so per job (there are ways of coalescing the bjobs calls if that's not cool).
+
+The procedure:
+
+- Pickle the class
+- Construct a bsub argument that runs a generic runner function with the path to the pickled class
+- Runner function loads the class from pickle
+- Runner function hits the work button on it
+
+"""
+
+LOGGER = logging.getLogger('luigi-interface')
+
+
+def track_job(job_id):
+    """
+    Tracking is done by requesting each job and then searching for whether the job
+    has one of the following states:
+    - "RUN",
+    - "PEND",
+    - "SSUSP",
+    - "EXIT"
+    based on the LSF documentation
+    """
+    cmd = "bjobs -noheader -o stat %d" % job_id
```

Comment:
Should you use new style .format()? Old style hasn't been deprecated yet, but the threat is still hanging out there.